src/Lean/Linter/Basic.lean

```diff
@@ -0,0 +1,124 @@
+import Lean.Elab.Command
+import Lean.Linter.Util
+import Lean.Elab.InfoTree
+import Lean.Server.InfoUtils
+import Lean.Server.References
+import Std.Data.HashMap
+
+namespace Lean.Linter
+open Lean.Server Std
+
+register_builtin_option linter.unusedVariables.nolint : Bool := {
+  defValue := false,
+  descr := "disable the 'unused variables' linter"
+}
+
+def unusedVariables : Linter := fun stx => do
+  let some stxRange := stx.getRange?
+    | pure ()
+
+  let fileMap := (← read).fileMap
+  let refs := dedupReferences <| combineFvars <| findReferences fileMap (← get).infoState.trees.toArray
+
+  let mut vars : HashMap FVarId (Option Reference × Array Reference) := .empty
```

Comment:
It's basically the same logic as in `RefInfo.addRef` and `ModuleRefs.addRef` combined with the splitting in `fvar`s and `const`s, but reimplementing this allows to keep the additional information in `Reference`. If we wanted to reuse `ModuleRefs`, we would need to include the additional information in `RefInfo`, which for now only includes ranges. This would make the `ToJson` instance a little more complicated, but I don't know how we would handle the `FromJson` conversion.